includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -901,4 +967,94 @@ public function update_order_status() {
 			wp_die();
 		}
 	}
+
+	/**
+	 * Add payment method via account screen.
+	 *
+	 * @throws Exception If payment method is missing.
+	 */
+	public function add_payment_method() {
+		try {
+			if ( ! isset( $_POST['wcpay-setup-intent'] ) ) { // phpcs:ignore WordPress.CSRF.NonceVerification.NoNonceVerification
+				throw new Exception( __( 'A WooCommerce Payments payment method was not provided', 'woocommerce-payments' ) );
+			}
+
+			$setup_intent_id = ! empty( $_POST['wcpay-setup-intent'] ) ? wc_clean( $_POST['wcpay-setup-intent'] ) : false; // phpcs:ignore WordPress.CSRF.NonceVerification.NoNonceVerification,WordPress.Security.ValidatedSanitizedInput.MissingUnslash
+
+			$customer_id = $this->customer_service->get_customer_id_by_user_id( get_current_user_id() );
+
+			if ( ! $setup_intent_id || null === $customer_id ) {
+				throw new Exception( __( "We're not able to add this payment method. Please try again later", 'woocommerce-payments' ) );
+			}
+
+			$setup_intent   = $this->payments_api_client->get_setup_intent( $setup_intent_id );
+			$payment_method = $setup_intent['payment_method'];
+
+			if ( 'succeeded' !== $setup_intent['status'] ) {
+				throw new Exception( __( 'Failed to add the provided payment method. Please try again later', 'woocommerce-payments' ) );
+			}
+
+			$payment_method_object = $this->payments_api_client->get_payment_method( $payment_method );
+			$this->token_service->add_token_to_user( $payment_method_object, wp_get_current_user() );
+
+			return [
+				'result'   => 'success',
+				'redirect' => wc_get_endpoint_url( 'payment-methods' ),
+			];
+		} catch ( Exception $e ) {
+			wc_add_notice( $e->getMessage(), 'error', [ 'icon' => 'error' ] );
+			Logger::log( 'Error when adding payment method: ' . $e->getMessage() );
+			return [
+				'result' => 'error',
+			];
+		}
+	}
+
+	/**
+	 * Create a setup intent when adding cards using the my account page.
+	 *
+	 * @throws Exception - When an error occurs in setup intent creation.
+	 */
+	public function create_setup_intent() {
+		$payment_method = $this->get_payment_method_from_request();
+
+		// Determine the customer adding the payment method, create one if we don't have one already.
+		$user        = wp_get_current_user();
+		$customer_id = $this->customer_service->get_customer_id_by_user_id( $user->ID );
+		if ( null === $customer_id ) {
+			$customer_id = $this->customer_service->create_customer_for_user( $user, "{$user->first_name} {$user->last_name}", $user->user_email );
```

Comment:
A user might not have `first_name` and `last_name` setup at this point, which would lead to customers having no name. Even though those details are updated on payment, it doesn't seem to update the payment method, which could lead to some transactions having a customer with the name "undefined undefined". Any payment method added through my account page before a customer name is defined will present this behavior.

What would be the best way to solve this? Should we choose one of the solutions below or is there a better way to do it?
- We could add a `name` field to the `add_payment_method` form in case the user has no customer created yet
- We could throw an error and ask the user to fill out his details using the `Account details` page before adding a payment method

Reply:
Yeah, it is, but while the customer is updated, the previously added payment methods are not. So the billing details are correct and the customer is updated, but still, the transaction and payment methods show the original customer name.

At least that was what I thought. I investigated this further and the "undefined undefined" name was caused on the creation of the payment method, by the checkout script.

That should be easy enough to fix by not passing the result of [`loadBillingDetails`](https://github.com/Automattic/woocommerce-payments/pull/758/files/d81aa9cd5cd30e126f41d2def3706f5b57990e4e#diff-abd459ff0f29b162345ad732a6a41279L158) when creating a payment method from the my account page since that page has no billing address fields.

This still leaves us with the problem that the payment method is created without billing details, so I'm thinking of updating the selected payment method with the data that's in the checkout form when processing a saved card payment. That way we don't end up with transactions without a billing address. What do you think?

Reply:
> So Stripe doesn't seem to care when processing payments. Do we? 

I don't think we do, especially if the behavior is the same in the Stripe extension. However, I get a blank customer name on the transactions when the payment method was added using my account page. Do you get it too?

![image](https://user-images.githubusercontent.com/7714042/87701481-7155cd80-c76e-11ea-9731-fde686b0aa07.png)

If we do not update the saved method billing details, we would be able to fix this missing customer data by using the `name` and `email` that we add to the metadata as a fallback for `billing_details` in the payments list and details screens. It would require a small change to the transaction caching, but it can be made as part of server/pull/296. I've implemented it in server/pull/296/de118eb and added the country to the metadata in https://github.com/Automattic/woocommerce-payments/pull/758/commits/7885f7aee6d2f03b0143e7e5c70fd8975d4f8f37. I also updated the payment details view to make use of them in https://github.com/Automattic/woocommerce-payments/pull/758/commits/41e65da3bc4200a4dab4f6527ac6ec34fab064da.

I went for the metadata approach because the customer fields are not expanded for `charge.succeeded` webhooks responses, but if you feel like that would be better we can always fetch the customer object when populating them.

> If we do, then maybe we could just update the name on the method used, not all methods tied to the user (my test user has about 135 payment methods saved with other extensions).

Absolutely! If we go that route we should just update the payment method being used in that checkout. I didn't make it that clear in the TLDR, sorry.

Reply:
Added the payment method update in https://github.com/Automattic/woocommerce-payments/commit/2f420e629ac810e615c067fa7dfc6bd4a3f49bc5.